sway/sway_text_node.c

```diff
@@ -0,0 +1,291 @@
+}
+
+static void handle_outputs_update(struct wl_listener *listener, void *data) {
+	struct text_buffer *buffer = wl_container_of(listener, buffer, outputs_update);
+	struct wlr_scene_outputs_update_event *event = data;
+
+	float scale = 0;
+	enum wl_output_subpixel subpixel = WL_OUTPUT_SUBPIXEL_UNKNOWN;
+
+	for (size_t i = 0; i < event->size; i++) {
+		struct wlr_scene_output *output = event->active[i];
+		if (subpixel == WL_OUTPUT_SUBPIXEL_UNKNOWN) {
+			subpixel = output->output->subpixel;
+		} else if (subpixel != output->output->subpixel) {
+			subpixel = WL_OUTPUT_SUBPIXEL_NONE;
+		}
+
+		if (scale != 0 && scale != output->output->scale) {
+			// drop down to gray scale if we encounter outputs with different
+			// scales or else we will have chromatic aberations
+			subpixel = WL_OUTPUT_SUBPIXEL_NONE;
+		}
+
+		if (scale < output->output->scale) {
+			scale = output->output->scale;
+		}
+	}
+
+	buffer->visible = event->size > 0;
+
+	if (scale != buffer->scale || subpixel != buffer->subpixel) {
+		buffer->scale = scale;
+		buffer->subpixel = subpixel;
+		render_backing_buffer(buffer);
+	}
+}
+
+static void handle_destroy(struct wl_listener *listener, void *data) {
+	struct text_buffer *buffer = wl_container_of(listener, buffer, destroy);
+
+	wl_list_remove(&buffer->outputs_update.link);
+	wl_list_remove(&buffer->destroy.link);
+
+	free(buffer->text);
+	free(buffer);
+}
+
+static void text_calc_size(struct text_buffer *buffer) {
+	struct sway_text_node *props = &buffer->props;
+
+	cairo_t *c = cairo_create(NULL);
+	cairo_set_antialias(c, CAIRO_ANTIALIAS_BEST);
+	get_text_size(c, config->font_description, &props->width, NULL,
+		&props->baseline, 1, props->pango_markup, "%s", buffer->text);
+	cairo_destroy(c);
+
+	wlr_scene_buffer_set_dest_size(buffer->buffer_node,
+		get_text_width(props), props->height);
+}
+
+struct sway_text_node *sway_text_node_create(struct wlr_scene_tree *parent,
+		char *text, float color[4], bool pango_markup) {
+	struct text_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (!buffer) {
+		return NULL;
+	}
+
+	struct wlr_scene_buffer *node = wlr_scene_buffer_create(parent, NULL);
+	if (!node) {
+		free(buffer);
+		return NULL;
+	}
+
+	buffer->buffer_node = node;
+	buffer->props.node = &node->node;
+	buffer->text = strdup(text);
+	if (!buffer->text) {
+		free(buffer);
+		wlr_scene_node_destroy(&node->node);
+		return NULL;
+	}
+
+	buffer->props.height = config->font_height;
+	buffer->props.pango_markup = pango_markup;
+	memcpy(&buffer->props.color, color, sizeof(float) * 4);
+
+	buffer->destroy.notify = handle_destroy;
+	wl_signal_add(&node->node.events.destroy, &buffer->destroy);
+	buffer->outputs_update.notify = handle_outputs_update;
+	wl_signal_add(&node->events.outputs_update, &buffer->outputs_update);
+
+	text_calc_size(buffer);
+
+	return &buffer->props;
+}
+
+void sway_text_node_set_color(struct sway_text_node *node, float color[4]) {
+	if (memcmp(&node->color, color, sizeof(float) * 4) == 0) {
```

Comment:
Can we use `sizeof(color)` here, and the other places we're doing `sizeof(float) * 4` manually?

Reply:
```
../sway/sway_text_node.c: In function ‘sway_text_node_set_color’:
../sway/sway_text_node.c:257:47: error: ‘sizeof’ on array function parameter ‘color’ will return size of ‘float *’ [-Werror=sizeof-array-argument]
  257 |         if (memcmp(&node->color, color, sizeof(color)) == 0) {
```

Nope, thanks C.

Reply:
Unfortunate.